modules/plugins/youtube.go

```diff
@@ -488,3 +664,567 @@ func (yt *YouTube) newMsg(content string) *discordgo.MessageSend {
+	if yt.feedsLoopRunning {
+		yt.logger().Error("youtube feeds loop already running")
+		return
+	}
+	yt.feedsLoopRunning = true
+
+	go yt.youtubeFeedsLoop()
+
+	return
+}
+
+func (yt *YouTube) youtubeFeedsLoop() {
+	defer helpers.Recover()
+	defer func() {
+		yt.Lock()
+		yt.feedsLoopRunning = false
+		yt.Unlock()
+
+		go func() {
+			yt.logger().Error("The checkYoutubeFeedsLoop died. Please investigate! Will be restarted in 60 seconds")
+			time.Sleep(60 * time.Second)
+
+			yt.Lock()
+			yt.runYoutubeFeedsLoop()
+			yt.Unlock()
+		}()
+	}()
+
+	if err := yt.quota.Init(yt); err != nil {
+		helpers.Relax(err)
+	}
+
+	for ; ; time.Sleep(10 * time.Second) {
+		err := yt.quota.UpdateCheckingInterval()
+		helpers.Relax(err)
+
+		yt.checkYoutubeFeeds()
+	}
+}
+
+func (yt *YouTube) checkYoutubeFeeds() {
+	yt.RLock()
+	defer yt.RUnlock()
+
+	t := time.Now().Unix()
+	entries, err := yt.readEntries(rethink.Row.Field("next_check_time").Le(t))
+	helpers.Relax(err)
+
+	for _, e := range entries {
+		switch e.ContentType {
+		case "channel":
+			e = yt.checkYoutubeChannelFeeds(e)
+		case "quota":
+			continue
+		default:
+			yt.logger().Error("unknown contents type: " + e.ContentType)
+		}
+
+		// update next check time
+		e = yt.setNextCheckTime(e)
+		err := yt.updateEntry(e)
+		helpers.Relax(err)
+	}
+}
+
+func (yt *YouTube) checkYoutubeChannelFeeds(e DB_Youtube_Entry) DB_Youtube_Entry {
+	c, err := e.getChannelContent()
+	if err != nil {
+		yt.logger().WithFields(logrus.Fields{
+			"id": e.ID,
+		}).Error(err)
+		return e
+	}
+
+	// set iso8601 time which will be used search query filter "published after"
+	lastSuccessfulCheckTime := time.Unix(e.LastSuccessfulCheckTime, 0)
+	publishedAfter := lastSuccessfulCheckTime.
+		Add(time.Duration(-1) * time.Hour).
+		Format("2006-01-02T15:04:05-0700")
+
+	// get updated feeds
+	feeds, err := yt.getChannelFeeds(c.ID, publishedAfter)
+	if err != nil {
+		yt.logger().Error("check channel feeds error: " + err.Error() + " channel name: " + c.Name + "id: " + e.ID)
+		return e
+	}
+
+	newPostedVideos := make([]string, 0)
+	alreadyPostedVideos := make([]string, 0)
+
+	// check if posted videos and post new videos
+	for i := len(feeds) - 1; i >= 0; i-- {
+		feed := feeds[i]
+
+		// only 'upload' type video can be posted
+		if feed.Snippet.Type != "upload" {
+			continue
+		}
+		videoId := feed.ContentDetails.Upload.VideoId
+
+		// check if the video is already posted
+		if yt.isPosted(videoId, c.PostedVideos) {
+			alreadyPostedVideos = append(alreadyPostedVideos, videoId)
+			continue
+		}
+
+		// make a message and send to discord channel
+		videoUrl := fmt.Sprintf(youtubeVideoBaseUrl, videoId)
+		msg := yt.newMsg(videoUrl)
+
+		_, err = cache.GetSession().ChannelMessageSendComplex(e.ChannelID, msg)
+		if err != nil {
+			yt.logger().Error(err)
+			break
+		}
+
+		newPostedVideos = append(newPostedVideos, videoId)
+
+		yt.logger().Info("Posting youtube video: " + feed.Snippet.Title)
```

Comment:
Please add the target channel ID to the log => `yt.logger().WithField("channelID", e.ChannelID).Info(…)`

Reply:
changed